nameresolution/consul/consul.go

```diff
@@ -0,0 +1,327 @@
+		return resolverCfg, err
+	}
+	resolverCfg.QueryOptions = getQueryOptionsConfig(cfg)
+
+	// if registering, set DaprPort in meta, needed for resolution
+	if resolverCfg.Registration != nil {
+		if resolverCfg.Registration.Meta == nil {
+			resolverCfg.Registration.Meta = map[string]string{}
+		}
+
+		resolverCfg.Registration.Meta[resolverCfg.DaprPortMetaKey] = daprPort
+	}
+
+	return resolverCfg, nil
+}
+
+func getClientConfig(cfg configSpec) *consul.Config {
+	// If no client config use library defaults
+	if cfg.ClientConfig != nil {
+		return cfg.ClientConfig
+	}
+
+	return consul.DefaultConfig()
+}
+
+func getRegistrationConfig(cfg configSpec, props map[string]string) (*consul.AgentServiceRegistration, error) {
+	// if advanced registration configured ignore other registration related configs
+	if cfg.AdvancedRegistration != nil {
+		return cfg.AdvancedRegistration, nil
+	} else if !cfg.SelfRegister {
+		return nil, nil
+	}
+
+	var appID string
+	var appPort string
+	var host string
+	var httpPort string
+	var ok bool
+
+	if appID, ok = props[nr.AppID]; !ok {
+		return nil, fmt.Errorf("nr metadata property missing: %s", nr.AppID)
+	}
+
+	if appPort, ok = props[nr.AppPort]; !ok {
+		return nil, fmt.Errorf("nr metadata property missing: %s", nr.AppPort)
+	}
+
+	if host, ok = props[nr.HostAddress]; !ok {
+		return nil, fmt.Errorf("nr metadata property missing: %s", nr.HostAddress)
+	}
+
+	if httpPort, ok = props[nr.DaprHTTPPort]; !ok {
+		return nil, fmt.Errorf("nr metadata property missing: %s", nr.DaprHTTPPort)
+	}
+
+	// if no health checks configured add dapr sidecar health check by default
+	if len(cfg.Checks) == 0 {
+		cfg.Checks = []*consul.AgentServiceCheck{
+			{
+				Name:     "Dapr Health Status",
+				CheckID:  fmt.Sprintf("daprHealth:%s", appID),
+				Interval: "15s",
+				HTTP:     fmt.Sprintf("http://%s:%s/v1.0/healthz", host, httpPort), // default assumes consul agent on svc host
+			},
+		}
+	}
+
+	appPortInt, err := strconv.Atoi(appPort)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing app port: %w", err)
+	}
+
+	return &consul.AgentServiceRegistration{
+		Name:    appID,
+		Address: host,
+		Port:    appPortInt,
+		Checks:  cfg.Checks,
+		Tags:    cfg.Tags,
+		Meta:    cfg.Meta,
+	}, nil
+}
+
+func getQueryOptionsConfig(cfg configSpec) *consul.QueryOptions {
+	// if no query options configured add default filter matching every tag in config
+	if cfg.QueryOptions == nil {
+		return &consul.QueryOptions{
+			UseCache: true,
+		}
+	}
+
+	return cfg.QueryOptions
+}
+
+func parseConfig(rawConfig interface{}) (configSpec, error) {
+	config := configSpec{}
+	rawConfig = convertGenericConfig(rawConfig)
+
+	data, err := json.Marshal(rawConfig)
+	if err != nil {
+		return config, fmt.Errorf("error serializing to json: %w", err)
+	}
+
+	decoder := json.NewDecoder(bytes.NewReader(data))
+	decoder.DisallowUnknownFields()
+
+	if err := decoder.Decode(&config); err != nil {
```

Comment:
It feels a little strange that this is being decoded into structs like `consul.QueryOptions` that have none fields like `ctx context.Context` which is not JSON friendly. My sense is to have separate structs the read in the config and then initialize the consul structs from there. WDYT? Is this JSON approach done elsewhere or in Consul itself?

Reply:
I get what you mean by strange, this did feel like a shortcut. The question here is, is it worth adding a mapping layer of almost identical types to avoid the unfriendly fields? My initial concern with this was more, if we upgraded the consul library version in the future then changes in the library could potentially break user configuration, do you agree?

Reply:
> My initial concern with this was more, if we upgraded the consul library version in the future then changes in the library could potentially break user configuration, do you agree?

Yes! Yes good point! I'll take compile time errors over ambiguous runtime bugs any day :)

Reply:
okay, this should be addressed in the latest commit

Reply:
Makes sense to me.  Please create an issue for tracking if you don't think you'll get to it as a fast follow to this PR.

Reply:
sorry there might be a bit of a misunderstanding - regarding what you said above _"My sense is to have separate structs the read in the config and then initialize the consul structs from there."_ this is the approach I've ultimately taken, so configuration is no longer tightly coupled to the consul api structs